ecp5/pack.cc

```diff
@@ -2706,6 +2721,86 @@ class Ecp5Packer
                     copy_constraint(ci, id_CLKI, id_CLKO, 1);
+                } else if (ci->type == id_DCSC) {
+                    if ((!ci->ports.count(id_CLK0) && !ci->ports.count(id_CLK1)) || !ci->ports.count(id_DCSOUT))
+                        continue;
+                    auto mode = str_or_default(ci->params, id_DCSMODE, "POS");
+
+                    // TODO: We can do this simplification if we know
+                    // MODESEL is tied to 0. I don't know how I can check this
+                    // if (mode == "CLK0_LOW" || mode == "CLK0_HIGH" || mode == "CLK0") {
+                    //     copy_constraint(ci, id_CLK0, id_DCSOUT, 1.0);
+                    //     continue;
+                    // } else if (mode == "CLK1_LOW" || mode == "CLK1_HIGH" || mode == "CLK1") {
+                    //     copy_constraint(ci, id_CLK1, id_DCSOUT, 1.0);
+                    //     continue;
+                    // } else if (mode == "LOW" || mode == "HIGH") {
+                    //     continue;
+                    // }
+
+                    std::unique_ptr<ClockConstraint> derived_constr = nullptr;
+                    std::vector<NetInfo*> in_ports = {
+                        ci->ports.at(id_CLK0).net,
+                        ci->ports.at(id_CLK1).net,
+                    };
+
+                    for (auto p : in_ports) { 
+                        if (p == nullptr || p->clkconstr == nullptr) {
+                            derived_constr = nullptr;
+                            break;
+                        }
+
+                        if (derived_constr == nullptr) {
+                            derived_constr = std::unique_ptr<ClockConstraint>(new ClockConstraint ());
+                            derived_constr->low = p->clkconstr->low;
+                            derived_constr->high = p->clkconstr->high;
+                            derived_constr->period = p->clkconstr->period;
+                            continue;
+                        }
+
+                        auto& c1 = p->clkconstr;
+                        auto& c2 = derived_constr;
+
+                        if (mode == "NEG") {
+                            derived_constr->low = DelayPair(
+                                std::min(c1->low.min_delay, c2->low.min_delay),
+                                std::max(
+                                    c1->low.max_delay + c2->period.max_delay,
+                                    c2->low.max_delay + c1->period.max_delay
+                                )
+                            );
+                        } else {
+                            derived_constr->low = DelayPair(
+                                std::min(c1->low.min_delay, c2->low.min_delay),
+                                std::max(c1->low.max_delay, c2->low.max_delay)
+                            );
+                        }
+
+                        if (mode == "POS") {
+                            derived_constr->high = DelayPair(
```

Comment:
@gatecat 

This maybe overconstrained too. At the transition point between clocks the worst case is that the clock will remain high or low for an additional period in the domain you are switching to. Since this only happens when transitioning it might be oke to remove it?

Reply:
No, I think this is a useful thing to have